File: built-in plugins dsp/mb-eq-parametric.cpp
/**
 * MB Parametric EQ
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 4-band parametric EQ
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_EQ_PARAMETRIC_H
#define MB_EQ_PARAMETRIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEqParametric : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-eq-parametric";
    static constexpr const char* PLUGIN_NAME    = "MB Parametric EQ";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float lowFreq = 80f;  // range [20, 500]
    float lowGain = 0f;  // range [-24, 24]
    float highFreq = 8000f;  // range [2000, 20000]
    float highGain = 0f;  // range [-24, 24]
    };

    MbEqParametric() = default;
    ~MbEqParametric() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.lowFreq = std::clamp(params.lowFreq, 20f, 500f);
        params.lowGain = std::clamp(params.lowGain, -24f, 24f);
        params.highFreq = std::clamp(params.highFreq, 2000f, 20000f);
        params.highGain = std::clamp(params.highGain, -24f, 24f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Parametric EQ
        return input;
    }
};

#endif // MB_EQ_PARAMETRIC_H

File: built-in plugins dsp/mb-eq-parametric.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbEqParametricPlugin: PluginDefinition = { id: 'mb-eq-parametric', slug: 'mb-eq-parametric', name: 'MB Parametric EQ', category: 'effect', type: 'eq', version: '1.0.0', description: '4-band parametric EQ', author: 'Max Booster', parameters: [{ id: 'lowFreq', name: 'Low Freq', type: 'float', defaultValue: 80, minValue: 20, maxValue: 500, automatable: true }, { id: 'lowGain', name: 'Low Gain', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, automatable: true }, { id: 'highFreq', name: 'High Freq', type: 'float', defaultValue: 8000, minValue: 2000, maxValue: 20000, automatable: true }, { id: 'highGain', name: 'High Gain', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, automatable: true }], defaultPreset: { lowFreq: 80, lowGain: 0, highFreq: 8000, highGain: 0 } };

export default MbEqParametricPlugin;
